bilal-waheed/blockchain/stacks-cli-tool

bilal-waheed opened the pull request:
## Overview

The PR contains the code for 3 different transactions:
1. stx token transfer
2. rbf transaction
3. dust transaction

## Files

1. tool.js file is supposed to be the entry point.
2. src folder contains separate files for all the transaction types
3. common.js contains some 'common' methods. ```getMempoolData``` is currently not being used anywhere.

bilal-waheed committed:
blockchain/cli-tool #stxTokenTransfer

WHAT: CLI tool for stx token transfer

WHY: To transfer tokens on stacks testnet

IMPACT: Low

bilal-waheed committed:
feat/cli-tool/transactions

WHAT: Add functionality for rbf and dust transactions

WHY: To allow user to make rbf and dust transactions using the CLI tool

IMPACT: low

bilal-waheed committed:
doc/stacks-cli-tool/readme

What: Add usage guide to README.md file

WHY: The guide contains the commands syntax to make transactions using the tool

Impact: Low

bilal-waheed committed:
feat/cli-tool/transactions

WHAT: Add support for sponsored transactions

WHY: To allow user to make sponsored transactions

IMPACT: low

bilal-waheed committed:
Revert "feat/cli-tool/transactions"

This reverts commit bf08e6586f629f134935b93fa933657c90e64cdc.

bilalanees98 commented on src/rbfTransaction.js:
```diff
@@ -0,0 +1,48 @@
+// Returns StacksTransaction object converted from raw transaction (bytes)
+const getRawTx = async (txid) => {
+  const res = await fetch(
+    `https://stacks-node-api.testnet.stacks.co/extended/v1/tx/${txid}/raw`
```
Configure to switch between testnet and mainnet on input from CLI

bilalanees98 commented on src/rbfTransaction.js:
```diff
@@ -0,0 +1,48 @@
+  const serializedTx = transaction.serialize().toString("hex");
+
+  //converting the string to Uint8Array
+  const serArr = Uint8Array.from(serializedTx.split(","));
```
Name not definitive. Maybe change to serializedArray?

bilalanees98 commented on src/rbfTransaction.js:
```diff
@@ -0,0 +1,48 @@
+  //converting the string to Uint8Array
+  const serArr = Uint8Array.from(serializedTx.split(","));
+
+  const response = await broadcastRawTransaction(
```
Create a helper func for broadcasting raw txs, will remove the need to add the api url repeatedly, in future. Will lower coupling and if you need to change anything in the broadcast function, you will have to change it in one place only

bilalanees98 commented on package.json:
```diff
@@ -0,0 +1,15 @@
+  "main": "transaction.js",
+  "type": "module",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
```
Add a run command. Will have to change the examples in the readme as well.
Something like => run: "node tool...."

bilalanees98 commented on src/common.js:
```diff
@@ -0,0 +1,58 @@
+
+const getMempoolData = async () => {
+  const mempoolNonceData = await fetch(
+    "https://stacks-node-api.testnet.stacks.co/extended/v1/address/ST1P44DKMKA01XPCKEAGF8BAGXZF8MRKY2SF20B11/nonces"
```
URL is not network agnostic

bilalanees98 commented on src/common.js:
```diff
@@ -0,0 +1,58 @@
+} from "@stacks/transactions";
+
+const calcNonce = async () => {
+  const network = process.argv[7];
```
A better idea is to create a constants.js file and gather all process.argv inputs there. Constants can be exported from there and imported wherever needed.

bilalanees98 reviewed (changes requested):
Please make the changes as specified in the comments.

bilal-waheed committed:
refactor: create constants.js file, create sendRawTransaction helper function, add npm script sto start the tool, convert urls to dynamic

what: code refactoring

why: to fix static url, to improve the code structure and make it easy-to-extend

impact: medium

bilal-waheed committed:
refactor: remove unnecessary comments

what: removal of comments

why: to improve code readability

impact: low

bilalanees98 reviewed (approved):
Update this PR with changes from the other one, and this should be good to go.
